File: app/src/test/java/com/doxigo/muchtoman/InstallmentsTest.kt
package com.doxigo.muchtoman

import android.content.Context
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/** «قسط»: the due-date arithmetic, the plan's figures, and which rows are offered as payments. */
class InstallmentsTest {

    private var n = 0

    /** The 31st of شهریور — the last month with a 31st, so the next due date has to clamp. */
    private val first = jalaliDay(1405, 6, 31)

    private fun entry(
        day: Long,
        rial: Long,
        direction: String = "out",
        owner: String = "",
    ): LedgerEntry {
        n++
        val ref = "s:%04d:0".format(n)
        return LedgerEntry(
            txn = Txn(
                ref = ref, srcHash = ref, seq = 0, at = tehranDayStart(day), day = day,
                bank = "SAMAN", accountId = "SAMAN", direction = direction,
                amountRial = rial, signedRial = if (direction == "out") -rial else rial,
                balanceRial = null, feeRial = null, mask = "", instrument = "unknown",
                merchant = "جایی", merchantNorm = "جایی", refNo = "", printedAt = "",
                channel = "unknown", unitPrinted = "none", inferred = false,
                parserVer = PARSER_VERSION,
            ),
            categoryId = "cat_shopping", categoryFa = "خرید", confidence = 95,
            needsReview = false, duplicate = false, transfer = false, ownerMemberId = owner,
        )
    }

    private fun plan(id: String = "phone", payment: Long = 10_000_000, count: Int = 3) = Goal(
        id = id, nameFa = "گوشی", targetRial = payment, kind = GoalKind.INSTALLMENT,
        period = GoalPeriod.MONTH, startsOn = first, endsOn = jalaliMonthsAfter(first, count - 1),
        createdAt = 0, updatedAt = 0,
    )

    private fun links(plan: Goal, vararg paid: Pair<String, Long>) =
        paid.associate { (ref, rial) -> ref to InstallmentLink(plan.id, rial) }

    @Test
    fun `a reminder comes once per due date, inside her days and waking hours`() {
        val p = plan()
        val second = jalaliMonthsAfter(first, 1)
        fun at(day: Long, hour: Int) = tehranDayStart(day) + hour * 3_600_000L
        val unpaid = installmentProgress(p, emptyMap(), emptyList(), first - 1)

        val news = installmentNews(listOf(unpaid), 1, emptyMap(), at(first - 1, 10))
        assertEquals(listOf(unpaid to first), news.due)
        assertEquals("گوشی: سررسید قسط فرداست", installmentReminderTitle(p, first, first - 1))
        // Said once: the next sweep that day, and the one on the due day itself, stay quiet.
        assertTrue(installmentNews(listOf(unpaid), 1, news.marks, at(first - 1, 16)).due.isEmpty())
        assertTrue(installmentNews(listOf(unpaid), 1, news.marks, at(first, 10)).due.isEmpty())
        assertEquals(news.marks, installmentNews(listOf(unpaid), 1, news.marks, at(first, 10)).marks)
        // Not at three in the morning, not two days out on «a day before», not when off.
        assertTrue(installmentNews(listOf(unpaid), 1, emptyMap(), at(first - 1, 3)).due.isEmpty())
        assertTrue(installmentNews(listOf(unpaid), 1, emptyMap(), at(first - 2, 10)).due.isEmpty())
        assertTrue(installmentNews(listOf(unpaid), -1, emptyMap(), at(first - 1, 10)).due.isEmpty())

        // Paid ahead and linked: nothing to remind of until the payment after it.
        val paid = installmentProgress(p, links(p, "a" to 10_000_000), emptyList(), first - 1)
        assertTrue(installmentNews(listOf(paid), 1, emptyMap(), at(first - 1, 10)).due.isEmpty())
        // Paid but never linked: the calendar still moves on to next month's date.
        val behind = installmentProgress(p, emptyMap(), emptyList(), second - 1)
        assertEquals(listOf(behind to second), installmentNews(listOf(behind), 1, news.marks, at(second - 1, 10)).due)
    }

    @Test
    fun `a due date keeps its day where the month has one and clamps where it does not`() {
        assertEquals(JalaliDate(1405, 7, 30), jalaliOf(jalaliMonthsAfter(first, 1)))
        assertEquals(JalaliDate(1405, 12, 29), jalaliOf(jalaliMonthsAfter(first, 6)))
        // Measured from the first due date each time, not from the last one — so the 31st comes
        // back the moment a month has one, instead of having been worn down to the 29th for good.
        assertEquals(JalaliDate(1406, 1, 31), jalaliOf(jalaliMonthsAfter(first, 7)))
        assertEquals(JalaliDate(1405, 5, 31), jalaliOf(jalaliMonthsAfter(first, -1)))
    }

    @Test
    fun `the first due date is the next such day, today included`() {
        val today = jalaliDay(1405, 7, 10)
        assertEquals(today, firstInstallmentDue(10, today))
        assertEquals(jalaliDay(1405, 7, 15), firstInstallmentDue(15, today))
        assertEquals(jalaliDay(1405, 8, 5), firstInstallmentDue(5, today))
        // مهر has no 31st: a plan due on the 31st is due on its last day, not in آبان.
        assertEquals(jalaliDay(1405, 7, 30), firstInstallmentDue(31, jalaliDay(1405, 7, 20)))
    }

    @Test
    fun `a plan is refused unless every answer is one, and is private when it is stored`() {
        val now = tehranDayStart(jalaliDay(1405, 7, 10)) + 1
        assertNull(newInstallment("x", "  ", 10_000_000, 12, 15, now))
        assertNull(newInstallment("x", "گوشی", 0, 12, 15, now))
        assertNull(newInstallment("x", "گوشی", MAX_PLAUSIBLE_RIAL + 1, 12, 15, now))
        assertNull(newInstallment("x", "گوشی", 10_000_000, 0, 15, now))
        assertNull(newInstallment("x", "گوشی", 10_000_000, MAX_INSTALLMENTS + 1, 15, now))
        assertNull(newInstallment("x", "گوشی", 10_000_000, 12, 0, now))
        assertNull(newInstallment("x", "گوشی", 10_000_000, 12, 32, now))

        val plan = newInstallment("x", " گوشی ", 10_000_000, 12, 15, now)!!
        assertEquals(GoalKind.INSTALLMENT, plan.kind)
        assertEquals("گوشی", plan.nameFa)
        // Never shared, so the sync — which publishes shared rows only — never sends it anywhere.
        assertFalse(plan.shared)
        assertEquals(jalaliDay(1405, 7, 15), plan.startsOn)
        assertEquals(jalaliDay(1406, 6, 15), plan.endsOn)
        assertEquals(12, installmentCount(plan))
    }

    @Test
    fun `partial payments settle one due payment between them`() {
        val p = plan()
        val progress = installmentProgress(p, links(p, "a" to 4_000_000, "b" to 6_000_000), emptyList(), first)
        assertEquals(10_000_000L, progress.paidRial)
        assertEquals(1, progress.paidCount)
        assertEquals(0L, progress.overdueRial)
        assertEquals(installmentDueOn(p, 1), progress.nextDue)
    }

    @Test
    fun `a payment is late only once its own day has passed`() {
        val p = plan()
        val paidOne = links(p, "a" to 10_000_000)
        val second = installmentDueOn(p, 1)
        // On the due day itself she can still pay it: not late yet.
        val dueToday = installmentProgress(p, paidOne, emptyList(), second)
        assertEquals(0L, dueToday.overdueRial)
        assertEquals(second, dueToday.nextDue)
        assertEquals(10_000_000L, installmentProgress(p, paidOne, emptyList(), second + 1).overdueRial)
    }

    @Test
    fun `a link too large for the plan reads as paid off, never as a negative balance`() {
        val p = plan()
        val huge = links(p, "a" to MAX_PLAUSIBLE_RIAL, "b" to MAX_PLAUSIBLE_RIAL)
        val progress = installmentProgress(p, huge, listOf(entry(first, 10_000_000)), first)
        assertEquals(30_000_000L, progress.paidRial)
        assertTrue(progress.done)
        assertEquals(1f, progress.share, 0f)
        assertNull(progress.nextDue)
        assertTrue(progress.candidates.isEmpty())
    }

    @Test
    fun `a payment still counts after the ledger has forgotten its message`() {
        // Sources past the thirteen-month horizon are pruned. A two-year loan must not start
        // reading as overdue in its fourteenth month for payments the app watched her make.
        val p = plan()
        val live = entry(first, 10_000_000)
        val progress = installmentProgress(
            p, links(p, live.txn.ref to 10_000_000, "s:pruned:0" to 10_000_000), listOf(live), first,
        )
        assertEquals(20_000_000L, progress.paidRial)
        assertEquals(listOf(live.txn.ref), progress.payments.map { it.txn.ref })
        assertEquals(10_000_000L, progress.olderRial)
    }

    @Test
    fun `payments offered are her own outgoing rows since the last due date, its exact amount first`() {
        val p = plan()
        val olderExact = entry(first - 5, 10_000_000)
        val newerOther = entry(first, 3_000_000)
        val newerExact = entry(first, 10_000_000)
        val incoming = entry(first, 10_000_000, direction = "in")
        val lastMonths = entry(jalaliMonthsAfter(first, -1), 10_000_000)
        val his = entry(first, 10_000_000, owner = "b".repeat(32))
        val transfer = entry(first, 10_000_000).copy(transfer = true)
        val paysOther = entry(first, 10_000_000)
        val progress = installmentProgress(
            p,
            mapOf(paysOther.txn.ref to InstallmentLink("fridge", 10_000_000)),
            listOf(olderExact, newerOther, newerExact, incoming, lastMonths, his, transfer, paysOther),
            first,
        )
        assertEquals(
            listOf(newerExact, olderExact, newerOther).map { it.txn.ref },
            progress.candidates.map { it.txn.ref },
        )
    }

    @Test
    fun `only her own outgoing payments can pay a plan, from either end of the link`() {
        assertTrue(installmentPayable(entry(first, 10_000_000), mineId = ""))
        assertFalse(installmentPayable(entry(first, 10_000_000, owner = "b".repeat(32)), mineId = ""))
        assertFalse(installmentPayable(entry(first, 10_000_000, direction = "in"), mineId = ""))
        assertFalse(installmentPayable(entry(first, 10_000_000).copy(transfer = true), mineId = ""))
        assertFalse(installmentPayable(entry(first, 10_000_000).copy(duplicate = true), mineId = ""))
    }

    @Test
    fun `a plan made from a payment starts on its day and counts it as the first installment`() {
        // Filed as قسط three days ago; the plan made from it must not push its schedule a month on.
        val paidOn = first - 3
        val plan = newInstallment("p", "گوشی", 10_000_000, 12, jalaliOf(paidOn).day, tehranDayStart(first), firstDue = paidOn)!!
        assertEquals(paidOn, plan.startsOn)
        assertEquals(12, installmentCount(plan))
        val payment = entry(paidOn, 10_000_000)
        val progress = installmentProgress(plan, links(plan, payment.txn.ref to 10_000_000), listOf(payment), first)
        assertEquals(1, progress.paidCount)
        assertEquals(0L, progress.overdueRial)
        assertEquals(installmentDueOn(plan, 1), progress.nextDue)
        assertEquals("p", installmentPaidBy(payment.txn.ref, listOf(progress))?.plan?.id)
        assertNull(installmentPaidBy("s:elsewhere:0", listOf(progress)))
    }

    @Test
    fun `only readable links to a plan that still exists are counted`() {
        val live = plan("phone")
        val gone = plan("tv").copy(deleted = true)
        fun decision(ref: String, value: String?, deleted: Boolean = false) =
            TxnDecision(ref, ref, DecisionKind.INSTALLMENT, value, 0, 0, deleted = deleted)
        val found = installmentLinks(
            listOf(
                decision("a", InstallmentLink("phone", 10_000_000).encode()),
                decision("b", InstallmentLink("tv", 10_000_000).encode()),
                decision("c", "phone:abc"),
                decision("d", "phone:-5"),
                decision("e", null),
                decision("f", InstallmentLink("phone", 10_000_000).encode(), deleted = true),
                TxnDecision("g", "g", DecisionKind.NOTE, "phone:10000000", 0, 0),
            ),
            listOf(live, gone),
        )
        assertEquals(mapOf("a" to InstallmentLink("phone", 10_000_000)), found)
    }
}

/** The one read the screen takes, against a real database: a plan is a plan and nothing else. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [28])
class InstallmentsLedgerTest {
    @Test
    fun `an installment is read as one, and never as a savings goal`() = runBlocking {
        val context = ApplicationProvider.getApplicationContext<Context>()
        val durable = Room.inMemoryDatabaseBuilder(context, DurableDb::class.java).build()
        val derived = Room.inMemoryDatabaseBuilder(context, DerivedDb::class.java).build()
        try {
            seedBuiltins(durable)
            val now = System.currentTimeMillis()
            val today = tehranDay(now)
            durable.goals().put(newInstallment("plan", "گوشی", 10_000_000, 3, jalaliOf(today).day, now)!!)
            durable.goals().put(
                Goal(
                    id = "trip", nameFa = "سفر", targetRial = 50_000_000, kind = GoalKind.SAVE,
                    period = GoalPeriod.ONCE, startsOn = today, createdAt = now, updatedAt = now,
                )
            )
            // Linked to a message the ledger no longer holds, which is the case the copied amount is for.
            durable.decisions().put(
                TxnDecision(
                    "d1", "s:gone:0", DecisionKind.INSTALLMENT,
                    InstallmentLink("plan", 10_000_000).encode(), now, now,
                )
            )

            val view = ledgerView(derived, durable)
            // Before, every row that was not a cap came back as a savings goal.
            assertEquals(listOf("trip"), view.goals.map { it.goal.id })
            assertTrue(view.budgets.isEmpty())
            val plan = view.installments.single()
            assertEquals("plan", plan.plan.id)
            assertEquals(10_000_000L, plan.paidRial)
            assertEquals(0L, plan.overdueRial)
        } finally {
            derived.close()
            durable.close()
        }
    }
}

File: pwa/test/installments.test.ts
/** InstallmentsTest.kt, literally: the due-date arithmetic, the plan's figures, and which rows are offered as payments. */
import { describe, expect, it } from 'vitest';
import { GoalKind, GoalPeriod } from '../src/goals';
import {
  MAX_INSTALLMENTS, encodeInstallmentLink, firstInstallmentDue, installmentCount, installmentDueOn,
  installmentLineFa, installmentLinks, installmentNews, installmentNoteFa, installmentPaidBy, installmentPayable,
  installmentProgress, installmentReminderBody, installmentReminderFa, installmentReminderTitle, newInstallment,
  tomanFieldToRial,
} from '../src/installments';
import type { InstallmentLink } from '../src/installments';
import { jalaliDay, jalaliMonthsAfter, jalaliOf, tehranDayStart } from '../src/jalali';
import type { Decision, Goal, LedgerEntry } from '../src/model';
import { MAX_PLAUSIBLE_RIAL } from '../src/sms';
import { entry as row, goalRow } from './plans-fixture';

/** The 31st of شهریور — the last month with a 31st, so the next due date has to clamp. */
const first = jalaliDay(1405, 6, 31);

const entry = (day: number, rial: number, direction: 'in' | 'out' = 'out', owner = ''): LedgerEntry =>
  row(day, direction === 'out' ? -rial : rial, { categoryId: 'cat_shopping', merchant: 'جایی', owner });

const plan = (id = 'phone', payment = 10_000_000, count = 3): Goal => goalRow({
  id, nameFa: 'گوشی', targetRial: payment, kind: GoalKind.INSTALLMENT, period: GoalPeriod.MONTH,
  startsOn: first, endsOn: jalaliMonthsAfter(first, count - 1),
});

const links = (p: Goal, ...paid: Array<[string, number]>): Map<string, InstallmentLink> =>
  new Map(paid.map(([ref, rial]) => [ref, { planId: p.id, rial }]));

describe('installments', () => {
  it('a reminder comes once per due date, inside her days and waking hours', () => {
    const p = plan();
    const second = jalaliMonthsAfter(first, 1);
    const at = (day: number, hour: number) => tehranDayStart(day) + hour * 3_600_000;
    const unpaid = installmentProgress(p, new Map(), [], first - 1);

    const news = installmentNews([unpaid], 1, {}, at(first - 1, 10));
    expect(news.due).toEqual([[unpaid, first]]);
    expect(installmentReminderTitle(p, first, first - 1)).toBe('گوشی: سررسید قسط فرداست');
    expect(installmentNews([unpaid], 1, news.marks, at(first - 1, 16)).due).toEqual([]);
    expect(installmentNews([unpaid], 1, news.marks, at(first, 10)).due).toEqual([]);
    expect(installmentNews([unpaid], 1, news.marks, at(first, 10)).marks).toEqual(news.marks);
    expect(installmentNews([unpaid], 1, {}, at(first - 1, 3)).due).toEqual([]);
    expect(installmentNews([unpaid], 1, {}, at(first - 2, 10)).due).toEqual([]);
    expect(installmentNews([unpaid], -1, {}, at(first - 1, 10)).due).toEqual([]);

    const paid = installmentProgress(p, links(p, ['a', 10_000_000]), [], first - 1);
    expect(installmentNews([paid], 1, {}, at(first - 1, 10)).due).toEqual([]);
    const behind = installmentProgress(p, new Map(), [], second - 1);
    expect(installmentNews([behind], 1, news.marks, at(second - 1, 10)).due).toEqual([[behind, second]]);
  });

  it('a due date keeps its day where the month has one and clamps where it does not', () => {
    expect(jalaliOf(jalaliMonthsAfter(first, 1))).toEqual({ year: 1405, month: 7, day: 30 });
    expect(jalaliOf(jalaliMonthsAfter(first, 6))).toEqual({ year: 1405, month: 12, day: 29 });
    expect(jalaliOf(jalaliMonthsAfter(first, 7))).toEqual({ year: 1406, month: 1, day: 31 });
    expect(jalaliOf(jalaliMonthsAfter(first, -1))).toEqual({ year: 1405, month: 5, day: 31 });
  });

  it('the first due date is the next such day, today included', () => {
    const today = jalaliDay(1405, 7, 10);
    expect(firstInstallmentDue(10, today)).toBe(today);
    expect(firstInstallmentDue(15, today)).toBe(jalaliDay(1405, 7, 15));
    expect(firstInstallmentDue(5, today)).toBe(jalaliDay(1405, 8, 5));
    expect(firstInstallmentDue(31, jalaliDay(1405, 7, 20))).toBe(jalaliDay(1405, 7, 30));
  });

  it('a plan is refused unless every answer is one, and is private when it is stored', () => {
    const now = tehranDayStart(jalaliDay(1405, 7, 10)) + 1;
    expect(newInstallment('x', '  ', 10_000_000, 12, 15, now)).toBeNull();
    expect(newInstallment('x', 'گوشی', 0, 12, 15, now)).toBeNull();
    expect(newInstallment('x', 'گوشی', MAX_PLAUSIBLE_RIAL + 1, 12, 15, now)).toBeNull();
    expect(newInstallment('x', 'گوشی', 10_000_000, 0, 15, now)).toBeNull();
    expect(newInstallment('x', 'گوشی', 10_000_000, MAX_INSTALLMENTS + 1, 15, now)).toBeNull();
    expect(newInstallment('x', 'گوشی', 10_000_000, 12, 0, now)).toBeNull();
    expect(newInstallment('x', 'گوشی', 10_000_000, 12, 32, now)).toBeNull();

    const p = newInstallment('x', ' گوشی ', 10_000_000, 12, 15, now)!;
    expect(p.kind).toBe(GoalKind.INSTALLMENT);
    expect(p.nameFa).toBe('گوشی');
    expect(p.shared).toBe(false);
    expect(p.startsOn).toBe(jalaliDay(1405, 7, 15));
    expect(p.endsOn).toBe(jalaliDay(1406, 6, 15));
    expect(installmentCount(p)).toBe(12);
  });

  it('partial payments settle one due payment between them', () => {
    const p = plan();
    const progress = installmentProgress(p, links(p, ['a', 4_000_000], ['b', 6_000_000]), [], first);
    expect(progress.paidRial).toBe(10_000_000);
    expect(progress.paidCount).toBe(1);
    expect(progress.overdueRial).toBe(0);
    expect(progress.nextDue).toBe(installmentDueOn(p, 1));
  });

  it('a payment is late only once its own day has passed', () => {
    const p = plan();
    const paidOne = links(p, ['a', 10_000_000]);
    const second = installmentDueOn(p, 1);
    const dueToday = installmentProgress(p, paidOne, [], second);
    expect(dueToday.overdueRial).toBe(0);
    expect(dueToday.nextDue).toBe(second);
    expect(installmentProgress(p, paidOne, [], second + 1).overdueRial).toBe(10_000_000);
  });

  it('a link too large for the plan reads as paid off, never as a negative balance', () => {
    const p = plan();
    const progress = installmentProgress(p, links(p, ['a', MAX_PLAUSIBLE_RIAL], ['b', MAX_PLAUSIBLE_RIAL]), [entry(first, 10_000_000)], first);
    expect(progress.paidRial).toBe(30_000_000);
    expect(progress.done).toBe(true);
    expect(progress.share).toBe(1);
    expect(progress.nextDue).toBeNull();
    expect(progress.candidates).toEqual([]);
  });

  it('a payment still counts after the ledger has forgotten its message', () => {
    const p = plan();
    const live = entry(first, 10_000_000);
    const progress = installmentProgress(p, links(p, [live.txn.ref, 10_000_000], ['s:pruned:0', 10_000_000]), [live], first);
    expect(progress.paidRial).toBe(20_000_000);
    expect(progress.payments.map((e) => e.txn.ref)).toEqual([live.txn.ref]);
    expect(progress.olderRial).toBe(10_000_000);
  });

  it('payments offered are her own outgoing rows since the last due date, its exact amount first', () => {
    const p = plan();
    const olderExact = entry(first - 5, 10_000_000);
    const newerOther = entry(first, 3_000_000);
    const newerExact = entry(first, 10_000_000);
    const incoming = entry(first, 10_000_000, 'in');
    const lastMonths = entry(jalaliMonthsAfter(first, -1), 10_000_000);
    const his = entry(first, 10_000_000, 'out', 'b'.repeat(32));
    const transfer = { ...entry(first, 10_000_000), transfer: true };
    const paysOther = entry(first, 10_000_000);
    const progress = installmentProgress(
      p,
      new Map([[paysOther.txn.ref, { planId: 'fridge', rial: 10_000_000 }]]),
      [olderExact, newerOther, newerExact, incoming, lastMonths, his, transfer, paysOther],
      first,
    );
    expect(progress.candidates.map((e) => e.txn.ref)).toEqual([newerExact, olderExact, newerOther].map((e) => e.txn.ref));
  });

  it('only her own outgoing payments can pay a plan, from either end of the link', () => {
    expect(installmentPayable(entry(first, 10_000_000), '')).toBe(true);
    expect(installmentPayable(entry(first, 10_000_000, 'out', 'b'.repeat(32)), '')).toBe(false);
    expect(installmentPayable(entry(first, 10_000_000, 'in'), '')).toBe(false);
    expect(installmentPayable({ ...entry(first, 10_000_000), transfer: true }, '')).toBe(false);
    expect(installmentPayable({ ...entry(first, 10_000_000), duplicate: true }, '')).toBe(false);
  });

  it('a plan made from a payment starts on its day and counts it as the first installment', () => {
    const paidOn = first - 3;
    const p = newInstallment('p', 'گوشی', 10_000_000, 12, jalaliOf(paidOn).day, tehranDayStart(first), '', paidOn)!;
    expect(p.startsOn).toBe(paidOn);
    expect(installmentCount(p)).toBe(12);
    const payment = entry(paidOn, 10_000_000);
    const progress = installmentProgress(p, links(p, [payment.txn.ref, 10_000_000]), [payment], first);
    expect(progress.paidCount).toBe(1);
    expect(progress.overdueRial).toBe(0);
    expect(progress.nextDue).toBe(installmentDueOn(p, 1));
    expect(installmentPaidBy(payment.txn.ref, [progress])?.plan.id).toBe('p');
    expect(installmentPaidBy('s:elsewhere:0', [progress])).toBeNull();
  });

  it('only readable links to a plan that still exists are counted', () => {
    const live = plan('phone');
    const gone = { ...plan('tv'), deleted: true };
    const decision = (ref: string, value: string | null, deleted = false, kind: Decision['kind'] = 'installment'): Decision =>
      ({ id: `${kind}:${ref}`, ref, kind, value, createdAt: 0, updatedAt: 0, deleted, memberId: '', familyRef: '' });
    const found = installmentLinks([
      decision('a', encodeInstallmentLink({ planId: 'phone', rial: 10_000_000 })),
      decision('b', encodeInstallmentLink({ planId: 'tv', rial: 10_000_000 })),
      decision('c', 'phone:abc'),
      decision('d', 'phone:-5'),
      decision('e', null),
      decision('f', encodeInstallmentLink({ planId: 'phone', rial: 10_000_000 }), true),
      decision('g', 'phone:10000000', false, 'note'),
    ], [live, gone]);
    expect(found).toEqual(new Map([['a', { planId: 'phone', rial: 10_000_000 }]]));
  });

  it("the plan's words, the reminder's and the setting's", () => {
    const p = plan();
    expect(installmentReminderTitle(p, first, first)).toBe('گوشی: سررسید قسط امروزه');
    expect(installmentReminderTitle(p, first, first - 2)).toBe('گوشی: سررسید قسط پس‌فرداست');
    expect(installmentReminderTitle(p, first, first - 3)).toBe('گوشی: سررسید قسط ۳ روز دیگه‌ست');
    expect(installmentReminderBody(p, first)).toBe('۱ میلیون تومان • ۳۱ شهریور ۱۴۰۵');
    expect([-1, 0, 1, 3].map(installmentReminderFa)).toEqual(['خاموش', 'همون روز', '۱ روز قبل', '۳ روز قبل']);

    const unpaid = installmentProgress(p, new Map(), [], first);
    expect(installmentLineFa(unpaid)).toBe('ماهی ۱ میلیون تومان • ۰ از ۳ قسط');
    expect(installmentNoteFa(unpaid, first)).toEqual(['سررسید قسط بعدی امروزه.', false]);
    expect(installmentNoteFa(installmentProgress(p, new Map(), [], first - 1), first - 1)).toEqual(['قسط بعدی: ۳۱ شهریور ۱۴۰۵', false]);
    expect(installmentNoteFa(installmentProgress(p, new Map(), [], first + 1), first + 1))
      .toEqual(['۱ میلیون تومان از قسط‌هایی که سررسیدشون گذشته، هنوز پرداخت نشده.', true]);
    expect(installmentNoteFa(installmentProgress(p, links(p, ['a', 30_000_000]), [], first), first)).toEqual(['همه‌ی قسط‌ها پرداخت شد.', true]);
  });

  it('a Toman field is whole Rial, or nothing this app would store', () => {
    expect(tomanFieldToRial('۱۲٬۵۰۰')).toBe(125_000);
    expect(tomanFieldToRial('0')).toBeNull();
    expect(tomanFieldToRial('abc')).toBeNull();
    expect(tomanFieldToRial(String(MAX_PLAUSIBLE_RIAL))).toBeNull();
  });
});
